sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,190 @@
+    >>> discretizer = Discretizer(n_bins=3, categorical_features=[2])
+    >>> discretizer.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    Discretizer(...)
+    >>> discretizer.cut_points_ # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    array([[ 2. , 3. , 4.3333...],
+           [ 4. , 5. , 4.6666...]])
+    >>> # Transforming X will move the categorical features to the last indices
+    >>> discretizer.transform(X)
+    array([[0, 0, 0, 0],
+           [2, 2, 2, 1]])
+    """
+    sparse_formats = ['csr', 'csc']
+
+    def __init__(self, n_bins=2, categorical_features=None):
+        self.n_bins = n_bins
+        self.categorical_features = categorical_features
+
+        # Attributes
+        self.min_ = None
+        self.max_ = None
+        self.cut_points_ = None
+        self.n_features_ = None
+        self.continuous_mask_ = None
+
+    def _set_continuous_mask(self):
+        """Sets a boolean array that determines which columns are
+        continuous features.
+        """
+        if self.categorical_features is None:
+            self.n_continuous_features_ = self.n_features_
+            self.continuous_mask_ = np.ones(self.n_features_, dtype=bool)
+            return
+
+        if len(self.categorical_features) > self.n_features_:
+            raise ValueError("The number of categorical indices is more than "
+                             "the number of features in the dataset.")
+
+        self.continuous_mask_ = np.ones(self.n_features_, dtype=bool)
+        self.continuous_mask_[self.categorical_features] = False
+
+        # Checks if there are duplicate columns from self.categorical_features,
+        # such as [-1, 0, 0, self.n_features - 1]
+        if self.n_features_ - len(self.categorical_features) \
+                != self.continuous_mask_.sum():
+            raise ValueError("Duplicate indices detected from input "
+                             "categorical indices.")
+
+        self.n_continuous_features_ = self.continuous_mask_.sum()
+
+    def fit(self, X, y=None):
+        """Finds the intervals of interest from the input data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse-matrix}, shape [n_samples, n_features]
+            The input data containing continuous features to be
+            discretized. Categorical columns are indicated by
+            categorical_columns from the constructor.
+
+        Returns
+        -------
+        self
+        """
+
+        if self.n_bins < 2:
+            raise ValueError("Discretizer received an invalid number "
+                             "of bins. Received {0}, expected at least 2"
+                             .format(self.n_bins))
+
+        X = check_array(X, accept_sparse=self.sparse_formats)
+        self.n_features_ = X.shape[1]
+
+        # Set the mask for the continuous features in the array
+        self._set_continuous_mask()
+
+        continuous = X[:, self.continuous_mask_]
```

Comment:
Slicing / masking w.r.t. features is only possible using CSC format. Please accept only CSC format in `check_array` above.